(Android) Upgrading CBL 1.x db. LiteCoreBridge.convertException

When CBL 2 automatically upgrades a 1.x db, all goes well but for the end, causing the entire upgrade to be rolled back (the new temp file is deleted by cbl.)
After cbl logs 'closing datafile' [having imported all documents correctly] it throws the following exception(s): 

```
I/LiteCore [DB]: Closing DataFile
W/System.err: CouchbaseLiteException{domain=2, code=18, msg=Cross-device link}
W/System.err:     at com.couchbase.lite.LiteCoreBridge.convertException(LiteCoreBridge.java:24)
W/System.err:     at com.couchbase.lite.Database.open(Database.java:888)
W/System.err:     at com.couchbase.lite.Database.<init>(Database.java:158)
...
W/System.err: Caused by: LiteCoreException{domain=2, code=18, msg=Cross-device link}
W/System.err:     at com.couchbase.litecore.C4Database.open(Native Method)
W/System.err:     at com.couchbase.litecore.C4Database.<init>(C4Database.java:21)
W/System.err:     at com.couchbase.lite.Database.open(Database.java:881)
W/System.err: 	... 16 more
```

---------------
As mentioned, the new db is not saved, and the whole thing repeats itself every time I query the 1.x db with cbl 2.
To be clear, ALL the documents are being logged as being imported ok. It's only afterwards when this happens, reversing the upgrade.
I have tried this on two physical Android 6 devices with different db's containing between 1500-3000 documents, with exactly the same results.

This, with DB22.
-nat

Issure is reproducible with following actual device.

Device: LGE Nexus 5 Android 6.0.1, API 23

```
03-20 13:53:01.378 18437-18437/? I/LiteCore: Upgrader upgrading db </storage/emulated/0/test/upgrade.cblite2/>; creating new db at </storage/emulated/0/test/upgrade.cblite2_TEMP/>
03-20 13:53:01.421 18437-18437/? I/LiteCore: Upgrading CBL 1.x database </storage/emulated/0/test/upgrade.cblite2/>, user_version=102)
03-20 13:53:01.422 18437-18437/? I/LiteCore: Importing doc 'doc_upgrade'
03-20 13:53:01.422 18437-18437/? I/LiteCore:         ...rev 2-05f7a7c9fb75865b041c1b441feee14e
03-20 13:53:01.422 18437-18437/? I/LiteCore:         ...rev 1-3fdf2109cb5287ee38b25f61d6d9d9cf
03-20 13:53:01.461 18437-18437/? E/UpgradeTest: Cross-device link
03-20 13:53:01.461 18437-18437/? W/System.err: CouchbaseLiteException{domain='POSIXErrorDomain', code=18, msg=Cross-device link}
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.lite.CBLStatus.convertException(CBLStatus.java:47)
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.lite.CBLStatus.convertException(CBLStatus.java:53)
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.lite.AbstractDatabase.open(AbstractDatabase.java:802)
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.lite.AbstractDatabase.<init>(AbstractDatabase.java:128)
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.lite.Database.<init>(Database.java:35)
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.upgradetest.MainActivity.databaseOperationSDCard(MainActivity.java:201)
03-20 13:53:01.462 18437-18437/? W/System.err:     at com.couchbase.upgradetest.MainActivity.onCreate(MainActivity.java:48)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.Activity.performCreate(Activity.java:6251)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.ActivityThread.-wrap11(ActivityThread.java)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.os.Looper.loop(Looper.java:148)
03-20 13:53:01.462 18437-18437/? W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:5417)
03-20 13:53:01.462 18437-18437/? W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
03-20 13:53:01.463 18437-18437/? W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
03-20 13:53:01.463 18437-18437/? W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
03-20 13:53:01.463 18437-18437/? W/System.err: Caused by: LiteCoreException{domain=2, code=18, msg=Cross-device link}
03-20 13:53:01.463 18437-18437/? W/System.err:     at com.couchbase.litecore.C4Database.open(Native Method)
03-20 13:53:01.463 18437-18437/? W/System.err:     at com.couchbase.litecore.C4Database.<init>(C4Database.java:38)
03-20 13:53:01.463 18437-18437/? W/System.err:     at com.couchbase.lite.AbstractDatabase.open(AbstractDatabase.java:800)
03-20 13:53:01.463 18437-18437/? W/System.err: 	... 16 more
```

--- a/android/CouchbaseLite/src/ce/java/com/couchbase/lite/DatabaseConfiguration.java
+++ b/android/CouchbaseLite/src/ce/java/com/couchbase/lite/DatabaseConfiguration.java
@@ -19,16 +19,21 @@ package com.couchbase.lite;
 
 import android.content.Context;
 
+import java.io.File;
+
 /**
  * Configuration for opening a database.
  */
 public final class DatabaseConfiguration {
+    private static final String TEMP_DIR_NAME = "tmp";
+
     //---------------------------------------------
     // member variables
     //---------------------------------------------
     private boolean readonly = false;
     private Context context = null;
     private String directory = null;
+    private boolean customDir = false;
 
     //---------------------------------------------
     // Constructors
@@ -40,6 +45,7 @@ public final class DatabaseConfiguration {
         this.readonly = false;
         this.context = context;
         this.directory = context.getFilesDir().getAbsolutePath();
+        this.customDir = false;
     }
 
     public DatabaseConfiguration(DatabaseConfiguration config) {
@@ -48,6 +54,7 @@ public final class DatabaseConfiguration {
         this.readonly = false;
         this.context = config.context;
         this.directory = config.directory;
+        this.customDir = config.customDir;
     }
 
     //---------------------------------------------
@@ -66,6 +73,7 @@ public final class DatabaseConfiguration {
         if (readonly)
             throw new IllegalStateException("DatabaseConfiguration is readonly mode.");
         this.directory = directory;
+        this.customDir = true;
         return this;
     }
 
@@ -93,6 +101,16 @@ public final class DatabaseConfiguration {
     }
 
     String getTempDir() {
-        return context.getCacheDir().getAbsolutePath();
+        if (customDir) {
+            File cache = new File(directory, TEMP_DIR_NAME);
+            if (!cache.exists())
+                if (!cache.mkdirs())
+                    return context.getCacheDir().getAbsolutePath();
+                else if (!cache.isDirectory())
+                    return context.getCacheDir().getAbsolutePath();
+            return cache.getAbsolutePath();
+        } else {
+            return context.getCacheDir().getAbsolutePath();
+        }
     }
 }

